libsolidity/ast/Types.cpp

```diff
@@ -362,6 +369,74 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
 	if (auto refType = dynamic_cast<ReferenceType const*>(&_type))
 		typeLocation = refType->location();
 
+	return usingForDirectives | ranges::views::filter([&](UsingForDirective const* _directive) -> bool {
+		// Convert both types to pointers for comparison to see if the `using for` directive applies.
+		// Note that at this point we don't yet know if the functions are actually usable with the type.
+		// `_type` may not be convertible to the function parameter type.
+		return
+			!_directive->typeName() ||
+			*TypeProvider::withLocationIfReference(typeLocation, &_type, true) ==
+			*TypeProvider::withLocationIfReference(
+				typeLocation,
+				_directive->typeName()->annotation().type,
+				true
+			);
+	}) | ranges::to<vector<UsingForDirective const*>>;
+}
+
+}
+
+Result<FunctionDefinition const*> Type::userDefinedOperator(Token _token, ASTNode const& _scope, bool _unaryOperation) const
+{
+	if (!typeDefinition())
+		return nullptr;
```

Comment:
I don't mean not checking it at all. Just not checking it here. I think you're calling this function in only one situation - when type checking an operator expression - and you surely already have a condition that catches this and reports an error. I think it would be better to have that check before the call and here just to be able to assert that the type has a definition.

The biggest problem I have with this is that this forces you to use `Result` and you also return `nullptr` for several different situations that the typechecker needs to discern anyway. I'd leave that to the type checker and keep the function simpler.